core/src/main/java/com/graphhopper/routing/util/spatialrules/CustomAreaIndex.java

```diff
@@ -0,0 +1,56 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.util.spatialrules;
+
+import org.locationtech.jts.geom.*;
+import org.locationtech.jts.index.strtree.STRtree;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CustomAreaIndex {
+    private final GeometryFactory gf;
+    private final STRtree index;
+
+    public CustomAreaIndex(List<CustomArea> customAreas) {
+        gf = new GeometryFactory();
+        index = new STRtree();
+        for (CustomArea customArea : customAreas) {
+            for (Polygon border : customArea.getBorders()) {
```

Comment:
We should use `PreparedPolygon` which is a lot faster for point-in-polygon checks

Reply:
I tried this and you are right it is much faster. I did a test with 100k random queries in Europe against an area index that uses the countries file and this took around 18s, but only like 0.2s when using prepared geometry :+1:
However, the filled lines optimization does not seem to make the queries faster and sometimes it is even slower. Therefore I would rather not include this, or is this especially useful for some corner cases (maybe smaller areas?)

Reply:
My test code is here if you want to try yourself: https://github.com/graphhopper/graphhopper/pull/2353/commits/436821e718c2db9e595d4c7dd1c3d7546a061fbb

Reply:
> However, the filled lines optimization does not seem to make the queries faster and sometimes it is even slower. Therefore I would rather not include this, or is this especially useful for some corner cases (maybe smaller areas?)

Interesting. If it's not really faster, i'd agree that we should remove it.

Reply:
In my tests it did not seem to be significant, maybe like 10%, but not sure to be honest. I'll remove it for now.